### PR TITLE
[RISCV] Add an experimental pseudoinstruction to represent a rematerializable constant materialization sequence.

### PR DESCRIPTION
**Pull request by topperc**
Rematerialization during register allocation is currently limited to a single instruction with no inputs.

This patch introduces a pseudoinstruction that represents the materialization of a constant. I've started with a sequence of 2 instructions for now, which covers at least the common LUI+ADDI(W) case. This instruction will be expanded into real instructions immediately after register allocation using a new pass. This gives the post-RA scheduler a chance to separate the 2 instructions to improve ILP.

I believe this matches the approach used by AArch64.

Unfortunately, this loses some CSE opportunies when an LUI value is used by multiple constants with different LSBs.

This feature is off by default and a new backend command line option is added to enable it for testing.

This avoids the spill and reloads reported in #69586.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVPostRAExpandPseudoInsts.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains a pass that expands the pseudo instruction pseudolisimm32
+// into target instructions. This pass should be run during the post-regalloc
+// passes, before assembly emission. It is used when the TunePseudoLISimm32
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVPostRAExpandPseudoInsts.cpp:11]: I don't see `TunePseudoLISimm32`.

File: llvm/test/CodeGen/RISCV/pr69586.ll [hunk 2]
```diff
@@ -0,0 +1,1980 @@
...
+; REMAT-NEXT:    addiw a0, a0, -1024
+; REMAT-NEXT:    add a0, a1, a0
+; REMAT-NEXT:    sf.vc.v.i 2, 0, v10, 0
+; REMAT-NEXT:    vse32.v v8, (a0)
+; REMAT-NEXT:    lui a0, 9
+; REMAT-NEXT:    addiw a0, a0, -512
+; REMAT-NEXT:    add a0, a1, a0
+; REMAT-NEXT:    sf.vc.v.i 2, 0, v8, 0
+; REMAT-NEXT:    vse32.v v10, (a0)
+; REMAT-NEXT:    lui a0, 9
```
Comment (michalt) [inline llvm/test/CodeGen/RISCV/pr69586.ll:1411]: Would there be a way of reducing the amount of redundant instructions? We generate a bunch of identical `lui` instructions and each time we need to follow that with a `addiw`.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 3]
```diff
@@ -1670,6 +1670,12 @@ def PseudoJump : Pseudo<(outs GPR:$rd), (ins pseudo_jump_symbol:$target), [],
                         "jump", "$target, $rd">,
                  Sched<[WriteIALU, WriteJalr, ReadJalr]>;
 
+// Pseudo for LI of simm32.
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:1673]: Can you add something to the comment to indicate this is currently purely for experimentation purposes?

File: llvm/lib/Target/RISCV/RISCVPostRAExpandPseudoInsts.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                                           MachineBasicBlock::iterator MBBI) {
+  DebugLoc DL = MBBI->getDebugLoc();
+
+  int64_t Val = MBBI->getOperand(1).getImm();
+
+  RISCVMatInt::InstSeq Seq = RISCVMatInt::generateInstSeq(
+      Val, MBB.getParent()->getSubtarget().getFeatureBits());
+  assert(!Seq.empty());
+
+  Register SrcReg = RISCV::X0;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVPostRAExpandPseudoInsts.cpp:95]: This looks very close to RISCVInstrInfo::movImm, any chance we can common code?

**Review by preames — APPROVED** [review]
LGTM w/optional suggestions

p.s. Forgot to say, my approval is for the experimental feature.  Whether this is the direction we want to move in overall is going to depend of more work (i.e. the CSE problem), and data collected once we've addressed obvious problems to see if this should be enabled by default.

**Pull request merged**